### PR TITLE
[ProductAttribute] Fix displaying attributes depending on locales and channels

### LINKED ISSUE #8220: RFC - Translatable atribute values behavior
Let's say we have a Sylius instance that operates in two languages:
 - en_US
 - fr_FR

And this shop is selling tShirts, that have 3 mandatory attributes:
 - fabric
 - max. degrees for washing
 - collar type

Fabric is a select attribute, that allows the following options:
 - cotton
 - linen
 - polyester

Max. degrees for washing is an integer attribute, while Collar type is a string attribute.

----

The current implementation for selectable attributes in Sylius allows the following configuration for selectable attributes:

image

As a business owner, 
I am not able to define in this screen the french translation of cotton or linen. 

The only way I can do that with the current format is put all values, no matter what's the locale and select the translation on product level. Example:

image

English attribute value for product:

image

French attribute value for product:
image

This solution is cumbersome and does not scale very well when you have more than 2 locales or attributes with more than 10 options for each locale. 

----

Another issue discovered with the current implementation is with integer or percentage attributes. Those attributes do not require a translation, as do string attributes.

Based on my initial example:
image

While it makes sense to add different translatable values on product for string attributes, for integer and percentage, it doesn't. See above that, as a content editor, you enter the value 100 twice.  This is not a thing you want to do when you have like... 6 locales.

or 4-5 integer attributes on a 6 locales shop.

----

My proposal is to define attribute values as translatable or not translatable. The followings must be translatable:
 - select
 - text
 - textarea
 
While the followings do not make any sense to be translatable:
 - checkbox
 - integer
 - percentage
 - date
 - datetime

Also, in my opinion, the select attribute values must be translatable, not the assignment on the product. Our custom implementation is like this:
Selectable attribute on product (en_US):
image
Same selectable attribute on product (ro_RO):
image

And for en_US.
image

The attribute options are configured in a screen like this:
image

While for string attributes it looks like this:

image

And for integer attributes like this:

image

----

My proposal with this RFC is to rework some of the current behavior in attributes until Sylius gets stable. I understood from @pjedrzejewski this implies BC Break and we should consult first, before any decision.

Looking forward to hearing your opinions.

### PR DESCRIPTION
**Pull request by GSadee**
| Q               | A
| --------------- | -----
| Branch?         | 1.0
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | partially fixes #8220
| License         | MIT

### CODE REVIEW

**Comment by GSadee** [comment]
@pamil I've added some specs for test this behaviour

File: src/Sylius/Component/Product/Model/Product.php [hunk 1]
```diff
@@ -440,22 +448,49 @@ protected function createTranslation(): ProductTranslationInterface
...
+     */
+    private function hasNotEmptyAttributeByCodeAndLocale(string $attributeCode, string $localeCode): bool
+    {
+        $attributeValue = $this->getAttributeByCodeAndLocale($attributeCode, $localeCode);
+
+        if (null === $attributeValue) {
+            return false;
+        }
+
+        if (empty($attributeValue->getValue())) {
```
Comment (pamil) [inline src/Sylius/Component/Product/Model/Product.php]: I guess we do not want it to be an `empty` call.

> The following things are considered to be empty:
>  - "" (an empty string)
>  - 0 (0 as an integer)
>  - 0.0 (0 as a float)
>  - "0" (0 as a string)
>  - NULL
>  - FALSE
>  - array() (an empty array)
>  - $var; (a variable declared, but without a value)

File: src/Sylius/Component/Product/spec/Model/ProductSpec.php [hunk 2]
```diff
@@ -120,6 +122,66 @@ function it_refuses_to_remove_non_product_attribute(AttributeValueInterface $att
...
+        $attributeValuePL->getCode()->willReturn('colour');
+        $attributeValuePL->getValue()->willReturn('Niebieski');
+
+        $this->addAttribute($attributeValueEN);
+        $this->addAttribute($attributeValuePL);
+
+        $this
+            ->getAttributesByLocale('pl_PL', 'en_US')
+            ->toArray()
+            ->shouldReturn([$attributeValuePL->getWrappedObject()])
```
Comment (pamil) [inline src/Sylius/Component/Product/spec/Model/ProductSpec.php]: Let's add another one which is testing the same call but returning the attribute from the fallback locale.

File: src/Sylius/Component/Product/spec/Model/ProductSpec.php [hunk 3]
```diff
@@ -120,6 +122,66 @@ function it_refuses_to_remove_non_product_attribute(AttributeValueInterface $att
...
+        $attributeValueFR->getCode()->willReturn('colour');
+        $attributeValueFR->getValue()->willReturn('Bleu');
+
+        $this->addAttribute($attributeValueEN);
+        $this->addAttribute($attributeValueFR);
+
+        $this
+            ->getAttributesByLocale('pl_PL', 'en_US', 'fr_FR')
+            ->toArray()
+            ->shouldReturn([$attributeValueEN->getWrappedObject()])
```
Comment (pamil) [inline src/Sylius/Component/Product/spec/Model/ProductSpec.php]: Let's add another ones which are testing the same call but:
 - returning the attribute from `$localeCode` locale
 - returning the attribute from `$baseLocaleCode` locale

File: src/Sylius/Component/Product/spec/Model/ProductSpec.php (hunk 2 above)
Comment (pamil) [inline src/Sylius/Component/Product/spec/Model/ProductSpec.php]: Btw. instead of `$this->getAttributesByLocale(...)->toArray()->shouldReturn([$mock->getWrappedObject()]);` you can use `$this->getAttributesByLocale(...)->shouldIterateAs([$mock]);`

File: src/Sylius/Component/Product/Model/Product.php (hunk 1 above)
Comment (pamil) [reply src/Sylius/Component/Product/Model/Product.php]: I'm afraid we're missing the tests for attributes that have values existing in base locale, but these values are considered empty (especially for booleans or integers).

Like with base locale `fr_FR` and current locale `pl_PL` and boolean attribute, what happens if:
 - boolean attribute has `false` value in `fr_FR` but `true` in `pl_PL`
 - integer attribute has `0` value in `fr_FR` but `1` in `pl_PL`

**Comment by pamil** [comment]
I've been thinking for some time whether to classify this as a bug fix or a new feature, but it looks like it is the latter. `getAttributeByCodeAndLocale` behaviour stays the same if the third parameter is not passed.

What do you think about targeting `master` branch with this PR? Some parts can be still merged to 1.0, like the first two Behat scenarios and specs that does not call that method with the thrid parameter, but it's not necessary.

**Comment by GSadee** [comment]
@pamil IMO this is a bug fix, because even if the behaviour of `getAttributeByCodeAndLocale` has not been changed, adding the third parameter in twig template on shop fixes the behaviour of displaying attributes there.

And these behat scenarios cannot be merge to `1.0` without changing the `getAttributeByCodeAndLocale`, because they will fail 💥

**Comment by pamil** [comment]
From my POV the current attributes behaviour is like for any other translatable resources: try to fetch a translation for current locale, if not found - try fallback locale. This PR adds a new feature, which is including also attributes in base locale.

As for Behat, the first two scenarios should pass, the other two are describing new features.

Anyway, it might be helpful to hear some other opinion than ours :)

File: features/product/viewing_products/viewing_product_attributes_in_different_locales_and_channels.feature [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        And I am browsing the channel "US Channel"
+
+    @ui
+    Scenario: Viewing a detailed page with product's attribute for current channel with its default locale
+        When I view product "T-shirt banana" in the "Polish (Poland)" locale
+        Then I should see the product attribute "T-shirt material" with value "Skórka banana"
+        And I should also see the product attribute "T-shirt details" with value "Banana is a very good material."
+
+    @ui
+    Scenario: Viewing a detailed page with product's attribute for current channel in different locale
```
Comment (lchrusciel) [inline features/product/viewing_products/viewing_product_attributes_in_different_locales_and_channels.feature]: The name of this scenario should have it is name changed with the scenario above. This one is using a default locale

File: src/Sylius/Component/Attribute/Model/AttributeSubjectInterface.php [hunk 5]
```diff
@@ -28,10 +28,15 @@ public function getAttributes(): Collection;
...
      * @param string $fallbackLocaleCode
+     * @param string|null $baseLocaleCode
      *
      * @return Collection|AttributeValueInterface[]
      */
-    public function getAttributesByLocale(string $localeCode, string $fallbackLocaleCode): Collection;
+    public function getAttributesByLocale(
+        string $localeCode,
+        string $fallbackLocaleCode,
+        ?string $baseLocaleCode = null
```
Comment (lchrusciel) [inline src/Sylius/Component/Attribute/Model/AttributeSubjectInterface.php]: Maybe we could start adding TODO messages with the information, that from 2.0 the third parameter would be mandatory?

**Comment by GSadee** [comment]
I couldn't 100% agree because this PR fixes the scenario: 
 - the base locale is `en_US`
 - the product has attribute with values in `en_US` and `pl_PL` locales 
 - this product is in channel with default locale `fr_FR`
 - in the shop, product is displaying in locale `pl_PL`

**Before:** the attribute woldn't be shown at all and it isn't a standard behaviour for translatable entities for me :)
**After:** the attribute would be shown in `pl_PL`

File: src/Sylius/Component/Attribute/Model/AttributeSubjectInterface.php (hunk 5 above)
Comment (pamil) [reply src/Sylius/Component/Attribute/Model/AttributeSubjectInterface.php]: If we intend to make it required then `E_USER_DEPRECATED` error should be triggered if value was not passed. To be honest, I'd rather make fallback locale code optional as well (there's no method to get product attributes values in given locale only now).

**Comment by pamil** [comment]
Thanks @GSadee for your patience! ☮️
